test/integration/cdi/example-driver/app/kubeletplugin.go

```diff
@@ -0,0 +1,207 @@
+
+	// create CDI directory if not exists
+	if err := os.MkdirAll(ex.config.cdiDir, os.FileMode(0750)); err != nil {
+		return fmt.Errorf("failed to create directory: %v", err)
+	}
+
+	// create DRA driver socket directory if not exists.
+	if err := os.MkdirAll(filepath.Dir(ex.config.draAddress), 0750); err != nil {
+		return fmt.Errorf("failed to create DRA driver socket directory: %v", err)
+	}
+
+	// Run a grpc server for the driver, that listens on draAddress
+	if err := plugin.StartNonblockingGrpcServer(ex.config.draAddress, &ex); err != nil {
+		return fmt.Errorf("failed to start a nonblocking grpc server: %v", err)
+	}
+
+	// Run a registration server and registers plugin with kubelet
+	if err := plugin.StartRegistrar(ex.config.driverName, ex.config.draAddress, ex.config.pluginRegistrationPath); err != nil {
+		return fmt.Errorf("failed to start registrar: %v", err)
+	}
+
+	return nil
+
+}
+
+func (ex *examplePlugin) NodePrepareResource(ctx context.Context, req *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
+
+	const (
+		deviceName = "exampledevice"
+		cdiVersion = "0.2.0"
+	)
+	kind := ex.config.driverName + "/" + deviceName
+
+	klog.Infof("NodePrepareResource is called: request: %+v", req)
+
+	klog.Infof("Creating CDI Files")
+	// resourceHandle in the form of map[string]string
+	var decodedResourceHandle map[string]string
+	if err := json.Unmarshal([]byte(req.ResourceHandle), &decodedResourceHandle); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal resourceHandle: %v", err)
+	}
+	// validate if resourceHandle
+	if err := ex.validateResourceHandle(decodedResourceHandle); err != nil {
+		return nil, fmt.Errorf("failed to validate resourceHandle: %v", err)
+	}
+	// create ContainerEdits in device spec and ContainerEdits
+	containerEditsInDeviceSpec := &cdiapi.ContainerEdits{ContainerEdits: &specs.ContainerEdits{}}
+	for key, val := range decodedResourceHandle {
+		containerEditsInDeviceSpec.Append(&cdiapi.ContainerEdits{ContainerEdits: &specs.ContainerEdits{
+			Env: []string{key + "=" + val},
+		}})
+	}
+	containerEdits := &cdiapi.ContainerEdits{ContainerEdits: &specs.ContainerEdits{}}
+
+	// create spec
+	spec, err := cdiapi.NewSpec(
+		&specs.Spec{
+			Version: cdiVersion,
+			Kind:    kind,
+			Devices: []specs.Device{{
+				Name:           deviceName,
+				ContainerEdits: *containerEditsInDeviceSpec.ContainerEdits,
+			}},
+			ContainerEdits: *containerEdits.ContainerEdits,
+		},
+		ex.getJsonFilePath(req.ClaimUid), 1)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new spec: %v", err)
+	}
+	klog.V(5).InfoS("CDI spec is validated")
+
+	// create bytes from spec, which would be written into the json file
+	jsonBytes, err := json.Marshal(*spec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal: %v", err)
+	}
+
+	// create json files
+	filePath := filepath.Dir(ex.getJsonFilePath(req.ClaimUid))
+	file, err := os.CreateTemp(filePath, filepath.Base(filePath))
```

Comment:
sorry for suggesting this. Please, follow @pohly suggestion to use deterministic file names.

Reply:
So, you mean,, go back to "writefile"?

Reply:
yes

Reply:
ok! Thanks.!